src/main/java/io/confluent/idesidecar/websocket/resources/WebsocketEndpoint.java

```diff
@@ -265,16 +197,101 @@ public void onMessage(String messageString, Session senderSession) {
     }
   }
 
+  /**
+   * Handle the first message from a workspace session, currently in purgatory,
+   * which should be a HELLO_WORKSPACE message.
+   *
+   * If it is a valid HELLO_WORKSPACE message, the workspace is authorized and added to the
+   * sessions map, and the new workspace count is broadcast to all workspaces (inclusive).
+   * Otherwise an error message is sent to the session and it is closed.
+   *
+   * The session is always removed from the purgatory set.
+   *
+   * @param session the session that sent the message.
+   * @param messageString the JSON string message.
+   */
+  private void handleHelloMessage(Session session, String messageString) {
+    // Regardless of if the message was good or bad, we remove the session from purgatory
+    // upon receipt of the first message.
+    purgatorySessions.remove(session);
```

Comment:
I'd recommend removing from purgatory later on, just before the session is added to the other map. Lots of things might go wrong here, and this method is not using try-with-finally blocks to ensure the session is either left as-is or both removed from the set and added to the map.

Reply:
Done.